@here/harp-mapview/lib/text/TextElementsRenderer.ts

```diff
@@ -280,13 +306,49 @@ export class TextElementsRenderer {
     /**
      * Reset internal state at the beginning of a frame.
      */
-    reset() {
+    reset(frameNumber: number) {
         this.m_screenCollisions.reset();
         for (const textRenderer of this.m_textRenderers) {
             textRenderer.textCanvas.clear();
             textRenderer.poiRenderer.reset();
         }
         this.m_initializedTextElementCount = 0;
+
+        const zoomLevel = Math.floor(this.m_mapView.zoomLevel);
+
+        // check that the level completely loaded
+        const zoomLevelCompletelyLoaded = this.m_mapView.visibleTileSet.dataSourceTileList.every(
+            source => {
+                const sourceZoomLevel = source.dataSource.getDisplayZoomLevel(zoomLevel);
+                return source.renderedTiles.every(
+                    tile => tile.tileKey.level === sourceZoomLevel || tile.tileKey.level === 0
+                );
+            }
+        );
+
+        if (zoomLevelCompletelyLoaded === false && this.m_renderStatesCleanedUp === true) {
+            this.m_renderStatesCleanedUp = false;
+        }
+
+        // if level do not loaded or old data cleaned up do nothing
+        if (!zoomLevelCompletelyLoaded || this.m_renderStatesCleanedUp) {
+            return;
+        }
+
+        // if level completely loaded clean up unnecessary states
+        for (let [key, textElementStates] of Object.entries(this.m_renderStateStore)) {
+            // states that older then 120 frames are removed
+            textElementStates = textElementStates.filter(
+                state => state.renderStateText.lastFrameNumber + 120 > frameNumber
```

Comment:
Why 120 is used? Saving this as constant with a descriptive name somewhere at the top and referencing here will be more readable.

Reply:
Yes please, say no to magic numbers :)